Sources/ClientRuntime/Networking/AwsChunked/AWSChunkedStream.swift

```diff
@@ -0,0 +1,159 @@
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import AwsCommonRuntimeKit
+
+/// Reads data from the input stream, chunks it, and passes it out through its output stream.
+class AWSChunkedStream {
+    private var inputStream: Stream
+    private var signingConfig: SigningConfig
+    private var previousSignature: String
+    private var trailingHeaders: Headers
+    var chunkedReader: AWSChunkedReader
+    private var outputStream = BufferedStream()
+
+    /// Actor used for ensuring stream reads are performed in order and one at a time.
+    private actor ReadCoordinator {
+        /// The most recent task performed by this actor.
+        private var task: Task<Data?, Error>?
+
+        /// Perform a new unit of work, after previously scheduled work has completed.
+        func add(_ block: @escaping () async throws -> Data?) async throws -> Data? {
+            let prevTask = self.task
+            let task = Task {
+                _ = await prevTask?.result
+                return try await block()
+            }
+            self.task = task
+            return try await task.value
+        }
+    }
+
+    /// The actor instance used by this AWSChunkedStream to enforce reading in series.
+    private let readCoordinator = ReadCoordinator()
+
+    init(
+        inputStream: Stream,
+        signingConfig: SigningConfig,
+        previousSignature: String,
+        trailingHeaders: Headers,
+        checksumAlgorithm: ChecksumAlgorithm? = nil
+    ) {
+        self.inputStream = inputStream
+        self.signingConfig = signingConfig
+        self.previousSignature = previousSignature
+        self.trailingHeaders = trailingHeaders
+        self.chunkedReader = AWSChunkedReader(
+            stream: self.inputStream,
+            signingConfig: self.signingConfig,
+            previousSignature: self.previousSignature,
+            trailingHeaders: self.trailingHeaders,
+            checksum: checksumAlgorithm
+        )
+    }
+}
+
+extension AWSChunkedStream: Stream {
+
+    /// Writes data to the input stream.
+    /// - Parameter data: The data to be written.
+    func write(contentsOf data: Data) throws {
+        try inputStream.write(contentsOf: data)
+    }
+
+    /// Closes the input stream.
+    ///
+    /// The output stream will close once the input stream closes, and all data from the
+    /// input stream has been chunked.
+    func close() {
+        inputStream.close()
+    }
+
+    func closeWithError(_ error: any Error) {
+        inputStream.closeWithError(error)
+    }
+
+    var position: Data.Index {
+        self.inputStream.position
+    }
+
+    /// Returns the length of the _input_ stream.
+    var length: Int? {
+        self.inputStream.length
+    }
+
+    var isEmpty: Bool {
+        self.inputStream.isEmpty
+    }
+
+    var isSeekable: Bool {
+        self.outputStream.isSeekable
+    }
+
+    func read(upToCount count: Int) throws -> Data? {
```

Comment:
read() and readAsync() will use the input stream and reader to create a new chunk once the output stream has been exhausted.  This will prevent chunks from building up in memory well in advance of when they can be transmitted.